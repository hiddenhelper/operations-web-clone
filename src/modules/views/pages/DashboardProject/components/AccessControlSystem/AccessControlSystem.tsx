import React, { memo, useEffect, useMemo } from 'react';
import { VictoryPie } from 'victory';

import Typography from '@material-ui/core/Typography';

import DashboardWidget from '../../../../shared/DashboardWidget';
import DashboardColumnSummary from '../../../../shared/DashboardWidget/DashboardColumnSummary';
import Skeleton from '../../../../shared/Skeleton';

import { AccessControlSystemModel, GeneralModel, StatisticsModel } from '../../../../../models';
import { useStyles as dashboardStyles } from '../../../Dashboard/styles';
import { useStyles } from '../../../DashboardWorker/styles';
import { getDefaultValue } from '../../../../../../utils/generalUtils';

export interface IAccessControlSystemProps {
  queryParams: GeneralModel.IQueryParams;
  acsWidget: StatisticsModel.IAcsSummaryStatistics;
  loading: GeneralModel.ILoadingStatus;
  fetchAcsWidget: (query: GeneralModel.IQueryParams) => void;
}

const AccessControlSystem = ({ queryParams, acsWidget, loading, fetchAcsWidget }: IAccessControlSystemProps) => {
  const dashboardGlobalClasses = dashboardStyles();
  const classes = useStyles();

  const chartData = useMemo(() => acsWidget || StatisticsModel.getAcsSummaryFallback(), [acsWidget]);
  const isLoading: boolean = useMemo(() => loading && loading.isLoading, [loading]);

  const pieData = useMemo(
    () => [
      { x: '', y: chartData.totalRecords, color: '#E5E5E5' },
      { x: '', y: chartData.totalFilter, color: '#006DF7' },
    ],
    [chartData]
  );

  const handheldTotal: number = useMemo(
    () => chartData.pieChart.find(item => item.segment === AccessControlSystemModel.AccessControlSystemType.HANDHELD)?.total,
    [chartData]
  );
  const enclosedTsTotal: number = useMemo(
    () => chartData.pieChart.find(item => item.segment === AccessControlSystemModel.AccessControlSystemType.ENCLOSED_TS)?.total,
    [chartData]
  );
  const openPortalTotal: number = useMemo(
    () => chartData.pieChart.find(item => item.segment === AccessControlSystemModel.AccessControlSystemType.PORTAL)?.total,
    [chartData]
  );
  const turnstileTotal: number = useMemo(
    () => chartData.pieChart.find(item => item.segment === AccessControlSystemModel.AccessControlSystemType.TURNSTILE)?.total,
    [chartData]
  );

  useEffect(() => {
    fetchAcsWidget(queryParams);
  }, [queryParams, fetchAcsWidget]);
  return (
    <DashboardWidget
      styleClass={`${dashboardGlobalClasses.smallWidgetMinHeight} ${dashboardGlobalClasses.widgetCenterContent}`}
      renderContent={() => (
        <div style={{ display: 'flex', height: '100%', justifyContent: 'flex-start', alignItems: 'center', paddingTop: '8px' }}>
          <Skeleton isLoading={isLoading} animation="wave" variant="text" width={126} height={126}>
            <div
              style={{
                width: '230px',
                height: '100%',
                marginRight: '24px',
                position: 'relative',
                display: 'flex',
                justifyContent: 'center',
                alignItems: 'center',
              }}
            >
              <VictoryPie data={pieData} innerRadius={120} labels={() => ''} colorScale={pieData.reduce((colors, entry) => [...colors, entry.color], [])} />
              <span
                style={{
                  fontSize: '38px',
                  position: 'absolute',
                  color: 'rgba(0, 0, 0, 0.7)',
                  display: 'flex',
                  flexDirection: 'column',
                  justifyContent: 'center',
                  alignItems: 'center',
                }}
              >
                {chartData.totalFilter}
                <span style={{ display: 'block', fontSize: '15px', marginTop: '-12px' }}>devices</span>
              </span>
            </div>
          </Skeleton>
          <DashboardColumnSummary
            title="Access Control System"
            subtitle={`From a total of ${chartData.totalRecords} devices`}
            isLoading={isLoading}
            renderActivity={
              <ul className={classes.columnActivityList}>
                <li className={classes.columnActivityListItem}>
                  <span>{getDefaultValue(handheldTotal, 0)}</span>
                  <Typography>Handheld</Typography>
                </li>
                <li className={classes.columnActivityListItem}>
                  <span>{getDefaultValue(enclosedTsTotal, 0)}</span>
                  <Typography>Enclosed TS</Typography>
                </li>
                <li className={classes.columnActivityListItem}>
                  <span>{getDefaultValue(openPortalTotal, 0)}</span>
                  <Typography>Open Portal</Typography>
                </li>
                <li className={classes.columnActivityListItem}>
                  <span>{getDefaultValue(turnstileTotal, 0)}</span>
                  <Typography>Turnstile</Typography>
                </li>
              </ul>
            }
          />
        </div>
      )}
    />
  );
};

export default memo(AccessControlSystem);
