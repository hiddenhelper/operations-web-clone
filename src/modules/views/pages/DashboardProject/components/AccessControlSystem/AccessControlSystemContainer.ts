import { Dispatch } from 'redux';
import { connect } from 'react-redux';

import { GeneralModel } from '../../../../../models';
import { GENERAL } from '../../../../../../constants';
import { IRootState } from '../../../../../state-mgmt/rootState';
import { statisticsState } from '../../../../../state-mgmt/statistics';

import AccessControlSystem from './AccessControlSystem';

export const mapStateToProps = (state: IRootState) => ({
  acsWidget: state.statistics.acsProjectStatistics,
  loading: state.general.loadingMap[GENERAL.LOADING_KEY.FETCH_PROJECTS_ACS_STATISTICS],
});

export const mapDispatchToProps = (dispatch: Dispatch) => ({
  fetchAcsWidget: (query: GeneralModel.IQueryParams) => dispatch(statisticsState.actions.fetchProjectAcsStart(query)),
});

export default connect(mapStateToProps, mapDispatchToProps)(AccessControlSystem);
