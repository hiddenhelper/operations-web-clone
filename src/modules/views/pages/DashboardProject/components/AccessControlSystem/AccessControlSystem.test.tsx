import React from 'react';
import { render, RenderResult } from '@testing-library/react';
import { MemoryRouter } from 'react-router';

import { getAccessControlSystemWidget_1 } from '../../../../../../test/entities';
import AccessControlSystem, { IAccessControlSystemProps } from './AccessControlSystem';

describe('AccessControlSystem Component', () => {
  global.console.warn = () => {
    /** */
  };
  let wrapper: RenderResult;
  let props: IAccessControlSystemProps;

  beforeEach(() => {
    const mockDate = jest.fn(() => new Date(Date.UTC(2020, 11, 14)).valueOf());
    Date.now = mockDate;
    props = {
      queryParams: { stateCode: '', period: 0 },
      acsWidget: getAccessControlSystemWidget_1(),
      loading: {
        isLoading: false,
        hasError: false,
        error: null,
      },
      fetchAcsWidget: jest.fn(),
    };
  });

  it('should render', () => {
    wrapper = render(
      <MemoryRouter>
        <AccessControlSystem {...props} />
      </MemoryRouter>
    );
    expect(wrapper.container).toMatchSnapshot();
  });

  it('should fetchAcsWidget', () => {
    props.acsWidget = null;
    render(
      <MemoryRouter>
        <AccessControlSystem {...props} />
      </MemoryRouter>
    );
    expect(props.fetchAcsWidget).toHaveBeenCalled();
  });

  it('should render loading', () => {
    props.loading = { isLoading: true, hasError: false, error: null };
    wrapper = render(
      <MemoryRouter>
        <AccessControlSystem {...props} />
      </MemoryRouter>
    );
    expect(wrapper.container).toMatchSnapshot();
  });
});
