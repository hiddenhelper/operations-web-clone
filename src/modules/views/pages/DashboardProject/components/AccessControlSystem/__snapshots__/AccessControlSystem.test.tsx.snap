// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessControlSystem Component should render 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-smallWidgetMinHeight-41 makeStyles-widgetCenterContent-43 makeStyles-widget-72 MuiPaper-elevation0 MuiPaper-rounded"
  >
    <div
      style="display: flex; height: 100%; justify-content: flex-start; align-items: center; padding-top: 8px;"
    >
      <div
        style="width: 230px; height: 100%; margin-right: 24px; position: relative; display: flex; justify-content: center; align-items: center;"
      >
        <div
          class="VictoryContainer"
          style="height: 100%; width: 100%; pointer-events: none; position: relative;"
        >
          <svg
            height="400"
            role="img"
            style="pointer-events: all; width: 100%; height: 100%;"
            viewBox="0 0 400 400"
            width="400"
          >
            <g>
              <path
                d="M9.184850993605149e-15,-150A150,150,0,1,1,-85.58535553847331,-123.18744626524699L-68.46828443077865,-98.5499570121976A120,120,0,1,0,7.34788079488412e-15,-120Z"
                role="presentation"
                shape-rendering="auto"
                style="fill: #E5E5E5; padding: 10px; stroke: transparent; stroke-width: 1;"
                transform="translate(200, 200)"
              />
              <path
                d="M-85.58535553847331,-123.18744626524699A150,150,0,0,1,-2.7554552980815446e-14,-150L-2.2043642384652355e-14,-120A120,120,0,0,0,-68.46828443077865,-98.5499570121976Z"
                role="presentation"
                shape-rendering="auto"
                style="fill: #006DF7; padding: 10px; stroke: transparent; stroke-width: 1;"
                transform="translate(200, 200)"
              />
              <text
                direction="inherit"
                dx="0"
                id="pie-labels-0"
                x="251"
                y="373.97"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="padding: 20px; font-family: 'Gill Sans', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; fill: #252525; stroke: transparent;"
                  text-anchor="middle"
                  x="251"
                />
              </text>
              <text
                direction="inherit"
                dx="0"
                id="pie-labels-1"
                x="149"
                y="35.97"
              >
                <tspan
                  dx="0"
                  dy="0"
                  style="padding: 20px; font-family: 'Gill Sans', 'Seravek', 'Trebuchet MS', sans-serif; font-size: 14px; letter-spacing: normal; fill: #252525; stroke: transparent;"
                  text-anchor="middle"
                  x="149"
                />
              </text>
            </g>
          </svg>
          <div
            style="z-index: 99; position: absolute; top: 0px; left: 0px; width: 100%; height: 100%;"
          >
            <svg
              height="400"
              style="overflow: visible; width: 100%; height: 100%;"
              viewBox="0 0 400 400"
              width="400"
            />
          </div>
        </div>
        <span
          style="font-size: 38px; position: absolute; color: rgba(0, 0, 0, 0.7); display: flex; flex-direction: column; justify-content: center; align-items: center;"
        >
          46
          <span
            style="display: block; font-size: 15px; margin-top: -12px;"
          >
            devices
          </span>
        </span>
      </div>
      <div
        class=" makeStyles-columnSummaryWrapper-108"
      >
        <p
          class="MuiTypography-root makeStyles-columnSummaryTitle-109 MuiTypography-body1"
        >
          Access Control System
        </p>
        <p
          class="MuiTypography-root makeStyles-columnSummarySubtitle-110 MuiTypography-body1"
        >
          From a total of 430 devices
        </p>
        <div
          class="makeStyles-columnSummaryActivity-111"
        >
          <ul
            class="makeStyles-columnActivityList-58"
          >
            <li
              class="makeStyles-columnActivityListItem-59"
            >
              <span>
                15
              </span>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Handheld
              </p>
            </li>
            <li
              class="makeStyles-columnActivityListItem-59"
            >
              <span>
                4
              </span>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Enclosed TS
              </p>
            </li>
            <li
              class="makeStyles-columnActivityListItem-59"
            >
              <span>
                15
              </span>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Open Portal
              </p>
            </li>
            <li
              class="makeStyles-columnActivityListItem-59"
            >
              <span>
                12
              </span>
              <p
                class="MuiTypography-root MuiTypography-body1"
              >
                Turnstile
              </p>
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AccessControlSystem Component should render loading 1`] = `
<div>
  <div
    class="MuiPaper-root makeStyles-smallWidgetMinHeight-265 makeStyles-widgetCenterContent-267 makeStyles-widget-296 MuiPaper-elevation0 MuiPaper-rounded"
  >
    <div
      style="display: flex; height: 100%; justify-content: flex-start; align-items: center; padding-top: 8px;"
    >
      <div
        class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
        style="width: 126px; height: 126px;"
      />
      <div
        class="MuiSkeleton-root MuiSkeleton-rect makeStyles-skeletonTextWrapper-336 MuiSkeleton-wave"
        style="width: 100%; height: 75px;"
      />
    </div>
  </div>
</div>
`;
